### PR TITLE
Intermediate PR: Interface Definition and PoC for Ninja Query

### PR DESCRIPTION
**Pull request by seongjaeny**
This PR is for receiving intermediate feedback. 

It includes the work on the interface definition as well as a preliminary proof-of-concept implementation located in the ninja-query folder. 
Please note that the tests are still in progress.

### CODE REVIEW

File: ninja_query/impl_poc.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+        Args:
+            query (str): The input query string.
+
+        Returns:
+            list[tuple[str, str]]: A list containing a single tuple of (COMMAND, ARGUMENT).
+                                    Returns an empty list if the query is empty.
+
+        Example:
+            "list_dependencies foo" -> [("LIST_DEPENDENCIES", "foo")]
```
Comment (kyotov) [inline ninja_query/impl_poc.py:86]: this kind of documentation belongs in the interface!

File: README.md [hunk 2]
```diff
@@ -1 +1,59 @@
...
\ No newline at end of file
+# Ninja Query Language
+
+This repository implements an interface for querying dependency graphs generated by Ninja builds. The goal is to provide a clean, well-documented public interface that is separated from internal implementation details.
+
+## Overview
+
+The project is divided into three main parts:
+
+- **Interface Package** (`ninja_query/interface`): Contains all abstract interface definitions such as `Dependency`, `BuildOutput`, `QueryParser`, `DependencyAnalyzer`, and `Builder`. This package defines what functionality is publicly available.
```
Comment (samanthagburek) [inline README.md:9]: I think these names are different than what is actually in the file, it might help to align the names.

File: README.md [hunk 3]
```diff
@@ -1 +1,59 @@
...
+├── factory.py          # Factory module for creating instances
+```
+
+
+## Usage Example
+
+To use the public API, simply import from the top-level package:
+
+```python
+from ninja_query import Dependency, BuildOutput, QueryParser, DependencyAnalyzer, Builder, get_parser, get_dependency_analyzer, get_builder
```
Comment (samanthagburek) [inline README.md:33]: I think these names don't align with what you have in the interface code, it might be helpful for them to match up.

File: ninja_query/interface/__init__.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+"""
+Public interface for the Ninja Query System.
+"""
+
+
+class Manifest:
```
Comment (samanthagburek) [inline ninja_query/interface/__init__.py:6]: I believe this what your README is referring to in the interface package, and this is where the names don't match up.

File: tests/test_manifest.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+"""Test cases for the Manifest interface."""
+
+import pytest
+from ninja_query import Manifest  # Adjust the import according to your project structure
+
+class TestManifest:
```
Comment (samanthagburek) [inline tests/test_manifest.py:6]: Could add more test cases for edge cases, but I know this pr is supposed to be just about the interface so this is just something to think about.

File: README.md [hunk 6]
```diff
@@ -1 +1,59 @@
...
+
+This repository implements an interface for querying dependency graphs generated by Ninja builds. The goal is to provide a clean, well-documented public interface that is separated from internal implementation details.
+
+## Overview
+
+The project is divided into three main parts:
+
+- **Interface Package** (`ninja_query/interface`): Contains all abstract interface definitions such as `Dependency`, `BuildOutput`, `QueryParser`, `DependencyAnalyzer`, and `Builder`. This package defines what functionality is publicly available.
+- **Implementation Package** (`ninja_query/impl`): Contains concrete implementations of the interfaces (e.g., `NinjaQueryParser`, `NinjaDependencyAnalyzer`, and `NinjaBuilder`).
+- **Factory Module** (`ninja_query/factory.py`): Provides dependency injection functions (`get_parser()`, `get_dependency_analyzer()`, `get_builder()`) to create instances of the concrete implementations. This module separates public interface creation from implementation details.
```
Comment (samanthagburek) [inline README.md:11]: I don't think this file has been created yet, but it might be helpful to create it and add docstrings for each of the functions just for explanations if you know what they will do.

File: README.md [hunk 7]
```diff
@@ -1 +1,59 @@
...
+# Ninja Query Language
+
+This repository implements an interface for querying dependency graphs generated by Ninja builds. The goal is to provide a clean, well-documented public interface that is separated from internal implementation details.
+
+## Overview
+
+The project is divided into three main parts:
+
+- **Interface Package** (`ninja_query/interface`): Contains all abstract interface definitions such as `Dependency`, `BuildOutput`, `QueryParser`, `DependencyAnalyzer`, and `Builder`. This package defines what functionality is publicly available.
+- **Implementation Package** (`ninja_query/impl`): Contains concrete implementations of the interfaces (e.g., `NinjaQueryParser`, `NinjaDependencyAnalyzer`, and `NinjaBuilder`).
```
Comment (samanthagburek) [inline README.md:10]: See comment below, same applies-it might be helpful to add the file.

File: tests/poc/ninja_query.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        dot_file = generate_ninja_dot(args.build_dir)
+        
+    try:
+        # Read the graph
+        depedency_analyzer.read_dot(dot_file)
+    except Exception as e:
+        print(f"Error processing dependency graph: {e}")
+
+    if args.list_dependencies: 
+        target = depedency_analyzer.get_target_id(args.list_dependencies)
```
Comment (samanthagburek) [inline tests/poc/ninja_query.py:191]: If get_target_id returns none, it could raise an error.

File: tests/test_manifest.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    """Test cases for the Manifest interface implementation."""
+
+    def test_deps(self):
+        """Test that deps returns the correct direct dependencies."""
+        class TestManifestImpl(Manifest):
```
Comment (samanthagburek) [inline tests/test_manifest.py:11]: Could use a pytest.fixture so you don't have to repeat this every test
